### ISSUE TITLE
Training gets stuck at the beginning

### ISSUE THREAD

**Issue by ErinYang96** [Kamlesh364/Mask-RCNN-TF2.7.0-keras2.7.0#5]
Hi,

My model seems to be stuck at the beginning of the training process. There is no error so I am unable to know why. I find that many people have encountered this problem on the official Mask_RCNN (see [#2168](https://github.com/matterport/Mask_RCNN/issues/2168), [#1899](https://github.com/matterport/Mask_RCNN/issues/1899), [#1962](https://github.com/matterport/Mask_RCNN/issues/1962)). 

They have mainly provided two solution:
1. set `workers = 0`, and `use_multiprocessing = False`, which I found that you have already done in the source file `model.py`;
2. update the GPU driver. My current GPU driver version is 512.15, which I suppose is quite new.

Some people say that the code can work fine in ubuntun system, but not on windows 10. Well, this is the same thing for me, but I haven't found any solution yet. By the way, my training dataset contains only 20 images, so it should not be so difficult to train it.

**The code that I run into problem is as follow, which is basically the same as the sample `train_shapes.ipny` provided by Matterport team.**
```
#### create Mask_RCNN model ####
model = modellib.MaskRCNN(mode='training', config=config, model_dir=MODEL_DIR) # Define the model

# Which weights to start with?
init_with = 'coco'  # coco, or last
ly_names=['mrcnn_class_logits', 'mrcnn_bbox_fc', 
'mrcnn_bbox', 'mrcnn_mask']
# ly_decodes=[unicode(s) for s in ly_names]

if init_with == 'coco':
    # %debug
     
    model.load_weights(COCO_MODEL_PATH, by_name=True,
                      exclude=ly_names)
elif init_with == 'last':
    model.load_weights(model.find_last(), by_name=True) # model.find_last()

#### Train the model layers ####
# Train the head layers
model.train(dataset_train, dataset_val,
            learning_rate=config.LEARNING_RATE, 
            epochs=1, # epoch=60 in original verison
            layers='heads',
            augmentation=augmentation)

# Fine-tune all layers
model.train(dataset_train, dataset_val, 
            learning_rate=config.LEARNING_RATE / 10,
            epochs=1, # epoch=140 in original version
            layers='all',
            augmentation=augmentation)

#### create a config instance used for inference ####
class InferenceConfig(Config):
    GPU_COUNT = 1
    IMAGES_PER_GPU = 1
    
    IMAGE_MAX_DIM = 1024
    IMAGE_MIN_DIM = 1024
    
    IMAGE_RESIZE_MODE = 'square'

inference_config = InferenceConfig()
batch_size = inference_config.BATCH_SIZE

#### load the last training weight for inference ####
model = modellib.MaskRCNN(mode="inference", 
                          config=inference_config,
                          model_dir=MODEL_DIR)

model.load_weights(model.find_last(), by_name=True) # 72
```

**When I run the above code, all I can get is this:**
```
Starting at epoch 0. LR=0.001

Checkpoint Path: C:\Mask_RCNN2.0\logs\trout20220520T1216\mask_rcnn_trout_{epoch:04d}.h5
Selecting layers to train
fpn_c5p5               (Conv2D)
fpn_c4p4               (Conv2D)
fpn_c3p3               (Conv2D)
fpn_c2p2               (Conv2D)
fpn_p5                 (Conv2D)
fpn_p2                 (Conv2D)
fpn_p3                 (Conv2D)
fpn_p4                 (Conv2D)
rpn_model              (Functional)
mrcnn_mask_conv1       (TimeDistributed)
mrcnn_mask_bn1         (TimeDistributed)
mrcnn_mask_conv2       (TimeDistributed)
mrcnn_mask_bn2         (TimeDistributed)
mrcnn_class_conv1      (TimeDistributed)
mrcnn_class_bn1        (TimeDistributed)
mrcnn_mask_conv3       (TimeDistributed)
mrcnn_mask_bn3         (TimeDistributed)
mrcnn_class_conv2      (TimeDistributed)
mrcnn_class_bn2        (TimeDistributed)
mrcnn_mask_conv4       (TimeDistributed)
mrcnn_mask_bn4         (TimeDistributed)
mrcnn_bbox_fc          (TimeDistributed)
mrcnn_mask_deconv      (TimeDistributed)
mrcnn_class_logits     (TimeDistributed)
mrcnn_mask             (TimeDistributed)
C:\Apps\Anaconda3\envs\CNNTest2\lib\site-packages\keras\optimizer_v2\gradient_descent.py:102: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)
```

**My environment setup is**
GPU: NVIDIA Quadro P620, 2GB (maybe it is too small for Mask_RCNN training??)
CUDA Driver API: 11.6
CUDA Runtime API: 11.2
cuDNN: 8.1
tensorflow 2.7.0
Keras 2.7.0 (changed due to the [previous problem](https://github.com/Kamlesh364/Mask-RCNN-TF2.7.0-keras2.8.0/issues/4))
OS: windows 10

Is there any version incompatible for my setup? Or some other problems? 
Many thanks if you can help me with it.

Best,
Erin

**Comment by Kamlesh364** [comment]
@ErinYang96 
This issue is quite common while working with low memory GPUs and not because of OS (either Ubuntu or Windows). I would recommend you to upgrade your GPU, but meanwhile you can finish your training by reducing, as much as possible, the training image size in `config.py` or training configuration (in case you have defined explicitly). 
please, do not hesitate to post again if you cannot make it.

best regards,
Kamlesh

**Comment by ErinYang96** [comment]
@Kamlesh364 I have now changed to our server with NVIDIA RTX A5000 (24G). And I have reduced the parameters in `config.py` as much as I can. But the problem remains the same. Or am I using the wrong version of CUDA and cuDNN, though it seems that I am connected to GPU by running `tf.config.experimental.list_physical_devices('GPU')`, which can produce the GPU number.
p.s. On the server, I have the same environment setup as I post above.

**Comment by Kamlesh364** [comment]
> @Kamlesh364 I have now changed to our server with NVIDIA RTX A5000 (24G). And I have reduced the parameters in `config.py` as much as I can. But the problem remains the same. Or am I using the wrong version of CUDA and cuDNN, though it seems that I am connected to GPU by running `tf.config.experimental.list_physical_devices('GPU')`, which can produce the GPU number. p.s. On the server, I have the same environment setup as I post above.

Please look into [this](https://stackoverflow.com/questions/52484793/keras-training-gets-stuck-in-lstm).

**Comment by ErinYang96** [comment]
> Please look into [this](https://stackoverflow.com/questions/52484793/keras-training-gets-stuck-in-lstm).

@Kamlesh364 I think the link you provide may not be the solution. For my case, my training just cannot start at all. Do you think [this](https://stackoverflow.com/questions/51176661/keras-seems-to-hang-after-call-to-fit-generator) may be the reason? But I don't know how to revise the source code based on this link due to my limited knowledge about deep learning. 
Also, some people say that the version of keras is the problem. Do you think it may work if I downgrade the keras version?
Many thanks if you can have a look into it.

**Comment by Kamlesh364** [comment]
> > Please look into [this](https://stackoverflow.com/questions/52484793/keras-training-gets-stuck-in-lstm).
> 
> @Kamlesh364 I think the link you provide may not be the solution. For my case, my training just cannot start at all. Do you think [this](https://stackoverflow.com/questions/51176661/keras-seems-to-hang-after-call-to-fit-generator) may be the reason? But I don't know how to revise the source code based on this link due to my limited knowledge about deep learning. Also, some people say that the version of keras is the problem. Do you think it may work if I downgrade the keras version? Many thanks if you can have a look into it.

In that case, please try with keras2.4.0 and let me know if you are able to solve this problem.

**Comment by ErinYang96** [comment]
@Kamlesh364 Hi, I try to replace keras=2.7.0 with 2.4.0 in Anaconda3. But maybe due to some package conflict, anaconda automatically converts my tensorflow version to 2.5.3, and h5py from 2.10.0 to 3.1. 0. I test with this new setup though, but the situation remains hanging at epoch 1/10. 
Also, I find that you have already implemented keras.utils.sequence for threadsafe. So this shouldn't be an issue about multiprocessing I suppose. 
Maybe it is still about version compatibility? Please tell me how you think about it.

**Comment by Kamlesh364** [comment]
> @Kamlesh364 Hi, I try to replace keras=2.7.0 with 2.4.0 in Anaconda3. But maybe due to some package conflict, anaconda automatically converts my tensorflow version to 2.5.3, and h5py from 2.10.0 to 3.1. 0. I test with this new setup though, but the situation remains hanging at epoch 1/10. Also, I find that you have already implemented keras.utils.sequence for threadsafe. So this shouldn't be an issue about multiprocessing I suppose. Maybe it is still about version compatibility? Please tell me how you think about it.

@ErinYang96 
I tried to find related resources to your issue and it seems to me that this issue can only be solved by tuning parameters in training configuration. I would recommend you to spend some time on StackOverflow (example: [1](https://stackoverflow.com/questions/48038417/keras-stops-working-on-first-epoch), [2](https://stackoverflow.com/questions/71181592/keras-model-stops-training-after-first-step-of-epoch)), keras and tensorflow official github repository (to find related issues).

Please spend some time on suggested platforms and let me know if you find anything. Meanwhile, I'll also try to replicate this issue on low memory GPUs and look for suitable solution.

`Note: Please make sure that you are looking for the stuff related TF2.`

best regards,
Kamlesh

**Comment by ErinYang96** [comment]
@Kamlesh364 Hi, I have figured out the reason. It turns out that I make a mistake when I load images at the previous step, which leads to lack of image ids. The code can train fine even with a low memory GPU after I fix this. Thanks for your support.

**Comment by Kamlesh364** [comment]
Glad to hear that you made it.
